src/ext/php7/engine_hooks.c

```diff
@@ -466,44 +408,54 @@ int ddtrace_wrap_fcall(zend_execute_data *execute_data) {
                            ZSTR_VAL(current_fbc->common.function_name));
         return vm_retval;
     }
+    /*
+    Internal functions are traced from the zend_execute_internal override for the sandbox API.
+    Ideally, we'd short-circuit early for internal functions, but since we have
+    to support the legacy API, we have to wait until after the dispatch hash table
+    lookup to determine if we can use zend_execute_internal (which only supports sandbox API).
+    */
+    if (current_fbc->type == ZEND_INTERNAL_FUNCTION && dispatch->options & DDTRACE_DISPATCH_POSTHOOK) {
+        return vm_retval;
+    }
     ddtrace_class_lookup_acquire(dispatch);  // protecting against dispatch being freed during php code execution
     dispatch->busy = 1;                      // guard against recursion, catching only topmost execution
 
     if (dispatch->options & DDTRACE_DISPATCH_POSTHOOK) {
-        _dd_trace_dispatch(dispatch, current_fbc, execute_data);
-    } else {
-        // Store original context for forwarding the call from userland
-        zend_function *previous_fbc = DDTRACE_G(original_context).fbc;
-        DDTRACE_G(original_context).fbc = current_fbc;
-        zend_function *previous_calling_fbc = DDTRACE_G(original_context).calling_fbc;
-
-        DDTRACE_G(original_context).calling_fbc = current_fbc->common.scope ? current_fbc : execute_data->func;
+        ddtrace_open_span(EX(call), dispatch);
+        return ZEND_USER_OPCODE_DISPATCH;
+    }
 
-        zval *this = _dd_this(EX(call));
+    // Store original context for forwarding the call from userland
+    zend_function *previous_fbc = DDTRACE_G(original_context).fbc;
+    DDTRACE_G(original_context).fbc = current_fbc;
+    zend_function *previous_calling_fbc = DDTRACE_G(original_context).calling_fbc;
```

Comment:
I didn't understand this despite the comment, sorry. Can you explain this more?

Reply:
This is ancient legacy code that stores parts of the execute_data for use with `dd_trace_forward_call()`. It's only green here because I removed the `else` block that surrounded it. :)